File: haskell/src/Solutions/Euler271.hs
module Solutions.Euler271 where
import Utils.NumberTheory(unPrime, primesUpTo)
import Utils.Numeric(fastPowerMod)

{-
Factorizing 13082761331670030, we can see that it is product of all primes up to 43.
By chinese remainder theorem, this congruence is equal to following congruences:
 x^3 = 1 mod 2
 x^3 = 1 mod 3
 ...
 x^3 = 1 mod 43
-}
primes = map unPrime $ primesUpTo 43 :: [Integer]
q = product primes

smallSolutions p = [x | x <- [1..p - 1], x^3 `mod` p == 1]
multiplier p = let m = q `div` p in m * fastPowerMod m (p - 2) p
inverse = map multiplier primes
compute = pred $ sum
        $ map (\solChoice -> sum (zipWith (*) solChoice inverse) `mod` q)
        $ mapM smallSolutions primes

euler271 :: IO String
euler271 = return $ show compute 

File: rust/src/solutions/euler271.rs
use crate::utils::numeric::modpow::ModPow;
use itertools::Itertools;
use std::ops::Sub;

/*
Factorizing 13082761331670030, we can see that it is product of all primes up to 43.
By chinese remainder theorem, this congruence is equal to following congruences:
 x^3 = 1 mod 2
 x^3 = 1 mod 3
 ...
 x^3 = 1 mod 43
*/

pub fn euler271() -> String {
    let primes = primal::Sieve::new(43)
        .primes_from(0)
        .take_while(|&p| p <= 43)
        .map(|p| p as u128)
        .collect::<Vec<_>>();
    let q: u128 = primes.iter().product();
    let inverses = primes
        .iter()
        .map(|&p| {
            let m = q / p;
            m * m.modpow(&(p - 2), &p)
        })
        .collect::<Vec<_>>();
    primes
        .into_iter()
        .map(|p| (1..p).filter(move |&x| (x * x * x) % p == 1))
        .multi_cartesian_product()
        .map(|sol_choice| {
            sol_choice
                .into_iter()
                .zip(&inverses)
                .map(|(x, &y)| x * y)
                .sum::<u128>()
                % q
        })
        .sum::<u128>()
        .sub(1)
        .to_string()
}
